src/unix/io_uring.c

```diff
@@ -0,0 +1,1340 @@
+            IOUR_REGION1_SIZE(iour));
+    iour_debug("rings %p, SQ array %p, CQEs %p, SQEs %p", iour->rings,
+               iour->sq_array, iour->cqes, iour->sqes);
+
+    iour_rings_init(iour);
+    spin_lock_init(&iour->lock);
+    iour->buf_count = iour->file_count = 0;
+    iour->bq = 0;
+    iour->eventfd = 0;
+    list_init(&iour->pollers);
+    list_init(&iour->timers);
+    iour->cq_timeouts = 0;
+    iour->noncancelable_ops = 0;
+    iour->shutdown = false;
+    iour->shutdown_completion = 0;
+    ret = allocate_fd(current->p, iour);
+    if (ret == INVALID_PHYSICAL) {
+        ret = -EMFILE;
+        goto err3;
+    }
+    iour_debug("fd %d", ret);
+    init_fdesc(h, &iour->f, FDESC_TYPE_IORING);
+    iour->f.close = init_closure(&iour->close, iour_close, iour);
+    params->features = IORING_FEAT_SINGLE_MMAP | IORING_FEAT_RW_CUR_POS;
+    params->sq_off.head = offsetof(io_rings, sq_head);
+    params->sq_off.tail = offsetof(io_rings, sq_tail);
+    params->sq_off.ring_mask = offsetof(io_rings, sq_mask);
+    params->sq_off.ring_entries = offsetof(io_rings, sq_entries);
+    params->sq_off.flags = offsetof(io_rings, sq_flags);
+    params->sq_off.dropped = offsetof(io_rings, sq_dropped);
+    params->sq_off.array = (u8 *)iour->sq_array - (u8 *)iour->rings;
+    runtime_memset((u8 *)params->sq_off.resv, 0, sizeof(params->sq_off.resv));
+    params->cq_off.head = offsetof(io_rings, cq_head);
+    params->cq_off.tail = offsetof(io_rings, cq_tail);
+    params->cq_off.ring_mask = offsetof(io_rings, cq_mask);
+    params->cq_off.ring_entries = offsetof(io_rings, cq_entries);
+    params->cq_off.overflow = offsetof(io_rings, cq_overflow);
+    params->cq_off.cqes = (u8 *)iour->cqes - (u8 *)iour->rings;
+    runtime_memset((u8 *)params->cq_off.resv, 0, sizeof(params->cq_off.resv));
+    return ret;
+err3:
+    deallocate(iour->vh, iour->user_rings, alloc_size);
+err2:
+    deallocate(h, iour->rings, alloc_size);
+err1:
+    deallocate(h, iour, sizeof(*iour));
+    return ret;
+}
+
+sysreturn io_uring_mmap(fdesc desc, u64 len, u64 mapflags, u64 offset)
+{
+    iour_debug("len %ld, flags 0x%x, offset 0x%x", len, mapflags, offset);
+    io_uring iour = (io_uring)desc;
+    u64 region_offset;
+    switch (offset) {
+    case IORING_OFF_SQ_RING:
+    case IORING_OFF_CQ_RING:
+        if (len > IOUR_REGION1_SIZE(iour))
+            return -EINVAL;
+        region_offset = 0;
+        break;
+    case IORING_OFF_SQES:
+        if (len > IOUR_REGION2_SIZE(iour))
+            return -EINVAL;
+        region_offset = IOUR_REGION1_SIZE(iour);
+        break;
+    default:
+        return -EINVAL;
+    }
+    u64 virt = u64_from_pointer(iour->user_rings) + region_offset;
+    map(virt, physical_from_virtual(iour->rings) + region_offset, len,
+        mapflags);
+    return virt;
+}
+
+simple_closure_function(1, 2, void, iour_efd_complete,
+                        u64, efd_val,
+                        thread, t, sysreturn, rv)
+{
+    closure_finish();
```

Comment:
Is this a placeholder function? Just trying to understand the closure_new / closure_get scheme ... where is the closure getting filled (and the function)?

Reply:
Here I'm just using the closure to piggyback the buffer (efd_val) passed to the eventfd write(). So instead of allocating the buffer and the completion closure separately, I'm doing everything with one allocation only. (Alternatively I could allocate just the buffer and use io_completion_ignore as completion.) The closure is being filled in the line `completion->efd_val = 1;`.

Reply:
Ok. I have at times thought it might be useful to be able to change the environment of a closure after it is allocated, prior to being applied, and these macros allow that to happen.